### DISCUSSION TITLE
Can I perform some custom logic on top of the generated code?

### DISCUSSION THREAD

**Discussion by NadChel** [mapstruct/mapstruct#3283] [Q&A]
Can I perform some custom logic on top of the generated code? Suppose I want something like this
```java
Post fromDtoAndAuthentication(PostDto dto, Authentication auth);
```
Mapstruct would do all the mapping and then invoke my custom function (kind of like an `@After` advice)
```java
this.setOwner((Account) auth.getPrincipal())
```
Is there any way I can do it?

**UPD:** I guess I should use `@AfterMapping`, though I don't fully understand how it works. Should I use the annotation `@Context` too for the parameter to be passed?
```java
@Mapper
public abstract class QuestionCommentMapper {
    @AfterMapping // ↓ should I annotate auth too?
    protected void enrichWithOwner(Authentication auth, @MappingTarget QuestionComment questionComment) {
        Account owner = AuthenticationProcessor.extractAccount(auth);
        questionComment.setOwner(owner);
    }

    public abstract QuestionComment dtoAndAuthenticationToQuestionComment(QuestionCommentRequestDto dto,
                                                                          @Context Authentication auth);
}
```

**Comment by filiphr** [comment]
The before / after mapping methods are meant to hook in custom logic that you want to execute. You can read more about them in [Mapping customization with before-mapping and after-mapping methods](https://mapstruct.org/documentation/stable/reference/html/#customizing-mappings-with-before-and-after).

Whether you should annotate the `Authentication` with `@Context` is entirely up to you. If you are not using the authentication as a source parameter then I would suggest to annotate it both in your mapping method and in the `@AfterMapping` method
